### PR TITLE
Users is satisfactory

### PR DESCRIPTION
**Pull request by charlesjhill**
What's up y'all; Users is now is now in a good enough place that' I'm willing to merge. Here's the new api routes and a brief description of what they do.

```
/api/v1/ <- API is now versioned 
      students/ <- our students are here 
      users/      <- ALL our users (students and admins) are here 
      rest-auth/ <- home to all things authentication 
            login/  <- You can login here. Posting to this with valid credentials yields a token 
            logout/ <- Post to here to logout 
            registration/ <- Post to here with some data to create a user; This will also yield a token 
```

Logging in currently requires an email, username, and password. I'm planning to fix this to allow either OR both.

Registration requires a username, email, password1, password2, a boolean specifying if the user is a student, and a m_number, which only matters if is_student is true.

Currently, none of our views are protected, so you don't require a token to access them. This is also intended to change.

QED

### CODE REVIEW

File: DjangoBackend/ucbearcatbandsproject/bands/serializers.py [hunk 1]
```diff
@@ -1,8 +1,11 @@
...
 
 
 class StudentSerializer(serializers.ModelSerializer):
+    user = UserSerializer(many=False, read_only=True)
+
     class Meta:
         model = Student
-        fields = ('id', 'first_name', 'last_name', 'm_number')
\ No newline at end of file
+        fields = ('m_number', 'user',)
```
Comment (benhollar) [inline DjangoBackend/ucbearcatbandsproject/bands/serializers.py]:
```suggestion
        fields = ('m_number', 'user',)
        
```

File: DjangoBackend/ucbearcatbandsproject/bands/urls.py [hunk 2]
```diff
@@ -7,7 +7,9 @@
...
 router.register(r'students', views.StudentViewSet)
 
-
 urlpatterns = [
-    path('', include(router.urls))
+    path('users/', include('ucbearcatbandsproject.users.urls')),
+    path('rest-auth/', include('rest_auth.urls')),
+    path('rest-auth/registration/', include('rest_auth.registration.urls')),
+    path('', include(router.urls)),
 ]
```
Comment (benhollar) [inline DjangoBackend/ucbearcatbandsproject/bands/urls.py]:
```suggestion
]

```

File: DjangoBackend/ucbearcatbandsproject/users/admin.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class CustomUserAdmin(UserAdmin):
+    add_form = CustomUserCreationForm
+    form = CustomUserChangeForm
+    model = CustomUser
+    list_display = ['email', 'username', 'is_student']
+
+
+admin.site.register(CustomUser, CustomUserAdmin)
```
Comment (benhollar) [inline DjangoBackend/ucbearcatbandsproject/users/admin.py]:
```suggestion
admin.site.register(CustomUser, CustomUserAdmin)

```

File: DjangoBackend/ucbearcatbandsproject/users/models.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from django.contrib.auth.models import AbstractUser
+from django.db import models
+
+# Create your models here.
+class CustomUser(AbstractUser):
+    is_student = models.BooleanField(default=True)
+
+    def __str__(self):
+        return self.email
```
Comment (benhollar) [inline DjangoBackend/ucbearcatbandsproject/users/models.py]:
```suggestion
        return self.email
        
```

File: DjangoBackend/ucbearcatbandsproject/users/serializers.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        # We should probably use Django Signals here? IDK
+        # Also, we should a let a client know they fucked up
+        # if they said they were a student but didn't provide an mnumber
+        if user.is_student and self.cleaned_data["m_number"]:
+            m_num = self.cleaned_data["m_number"]
+            stud = Student(m_number=m_num, user=user)
+            stud.save()
+        self.custom_signup(request, user)
+        setup_user_email(request, user, [])
+        return user
```
Comment (benhollar) [inline DjangoBackend/ucbearcatbandsproject/users/serializers.py]:
```suggestion
        return user
        
```

**Review by benhollar — CHANGES_REQUESTED** [review]
Hey fix the newlines

**Review by benhollar — APPROVED** [review]

**Comment by charlesjhill** [comment]
:thumbsup:

**Pull request merged**